lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -276,536 +344,505 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+      userDataValue = Util::Abi::UserDataMapping::StreamOutTable;
+    switch (m_shaderStage) {
+    case ShaderStageVertex:
+      userDataArgs.push_back(
+          UserDataArg(builder.getInt32Ty(), userDataValue, &intfData->entryArgIdxs.vs.streamOutData.tablePtr));
+      break;
+    case ShaderStageTessEval:
+      userDataArgs.push_back(
+          UserDataArg(builder.getInt32Ty(), userDataValue, &intfData->entryArgIdxs.tes.streamOutData.tablePtr));
+      break;
+    // Allocate dummy stream-out register for Geometry shader
+    case ShaderStageGeometry:
+      userDataArgs.push_back(UserDataArg(builder.getInt32Ty()));
+      break;
+    default:
+      llvm_unreachable("Should never be called!");
+      break;
     }
   }
+}
 
-  // Descriptor table and vertex buffer table
-  unsigned actualAvailUserDataCount = 0;
-  for (unsigned i = 0; i < userDataNodes.size(); ++i) {
-    auto node = &userDataNodes[i];
+// =====================================================================================================================
+// Add a UserDataArg to the vector for each user data node needed in user data SGPRs.
+//
+// This function does not check if it needs to spill because of running out of SGPRs. However it does
+// check if a node needs to be loaded from the spill table anyway because of its type, and in that case marks
+// it as "mustSpill".
+//
+// This function handles CS "fixed layout" by inserting an extra padding arg before a real user data arg if
+// necessary.
+//
+// @param userDataArgs : Vector to add args to
+// @param builder : IRBuilder to get types from
+void PatchEntryPointMutate::addUserDataArgs(SmallVectorImpl<UserDataArg> &userDataArgs, IRBuilder<> &builder) const {
 
-    // "IndirectUserDataVaPtr" can't be spilled, it is treated as internal user data
-    if (node->type == ResourceNodeType::IndirectUserDataVaPtr)
-      continue;
+  // Add entries from the root user data layout (not vertex buffer or streamout, and not unused ones).
+  auto intfData = m_pipelineState->getShaderInterfaceData(m_shaderStage);
+  bool useFixedLayout = m_shaderStage == ShaderStageCompute;
 
-    if (node->type == ResourceNodeType::StreamOutTableVaPtr)
-      continue;
+  // userDataIdx counts how many SGPRs we have already added, so we can tell if we need padding before
+  // an arg in a compute shader. We initialize with userDataArgs.size(), relying on the fact that all the
+  // ones added so far (global table, per-shader table, streamout table for graphics) use one register each.
+  unsigned userDataIdx = userDataArgs.size();
 
-    if (!isResourceNodeActive(node, true))
+  for (unsigned i = 0; i != m_pipelineState->getUserDataNodes().size(); ++i) {
+    const ResourceNode &node = m_pipelineState->getUserDataNodes()[i];
+    if (node.type == ResourceNodeType::IndirectUserDataVaPtr || node.type == ResourceNodeType::StreamOutTableVaPtr)
+      continue;
+    if (!isResourceNodeActive(&node, true))
       continue;
```

Comment:
Should this isResourceNodeActive check only be there for graphics shaders? Thinking of compute shaders and the fact that they don't allow remapping of user data based on PAL metadata...

Reply:
I think that would have the effect that an unused node appears in PAL metadata, rather than there just being a gap in user data sgprs in PAL metadata. I'm not sure if that has a bad effect (is dispatching a wave with s0,s1,s2,s3 any slower than s0,s3?), but it is a change in behavior from the old code.

The padding for compute is handled separately in a way that does not add metadata entries.

Reply:
Ah, I think I see the point.

From a HW perspective, gaps in the user data SGPRs don't help with wave launch performance: the HW always initializes a prefix of SGPRs.

Having the unused nodes could potentially hurt CPU performance in PAL though, I don't know if those are filtered out. Keeping the behavior as-is sounds good.